teps/0123-specifying-on-demand-retry-in-pipelinetask.md

```diff
@@ -0,0 +1,119 @@
+---
+status: proposed 
+title: Specifying on-demand-retry in a PipelineTask
+creation-date: '2022-09-16'
+last-updated: '2022-09-16'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0123: Specifying on-demand-retry in a PipelineTask
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+    - [CD Use Case](#cd-use-case)
+    - [CI Use Case](#ci-use-case)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a mechanism to allow the users to choose when to retry a failed `pipelineTask` of a pipeline.
+This kind of on demand retry of a `pipelineTask` is executed from scratch with the original `params` and `specifications`.
+The on demand retry of a `pipelineTask` allow the users to continue to retry until the task succeeds. After the offending task
+succeeds, the rest of the `pipeline` continues executing as usual until the completion.
+
+
+## Motivation
+
+The primary motivation of this proposal is to support pipeline authors to overcome any flakes and to allow running
+the same instance of the pipeline until it finishes.
+
+### Goals
+
+* Provide a mechanism to enable on-demand-retry of a `pipelineTask` in a `pipeline`.
+* Provide a mechanism to signal a failed `taskRun` to start executing with the original specifications. This mechanism
+will only be supported for a `pipelineTask` which opts in for on-demand-retry.
+* The resolved specifications (`params`, the resolved results from a parent task, `taskRef` or `taskSpec`, etc ) cannot
+be updated during such retry.
+* A pipeline author can specify more than one on-demand-retry `pipelineTasks` and at any level in the `DAG` including
+root node and leaf node.
+* On-demand-retry `pipelineTask` fails again after signalling a `taskRun` to start executing, it goes back into the same
+mode of on-demand-retry. The users have an option to retry such task again.
+* If on-demand-retry `pipelineTask` does not succeed until the pipeline hits the timeout, the failed tasks will be
```

Comment:
How would timeout work in such cases? 
Does the timeout clock continue running even if human intervention is needed?

It sounds like the `Pipeline` would be in a suspended kind of state whenever at least a `PipelineTask` that opted-in for `on-demand-retry` fails. 
Would the user be able to cancel a pipeline in such state?
Would the user be able to let the pipeline fail if they decided to?
What happens is resources (like the `Pod`) are pruned while waiting for an user to request a retry?

Reply:
> How would timeout work in such cases?
Does the timeout clock continue running even if human intervention is needed?

The timeout would work the same - no change from the timeout perspective. The timeout clock continues running and if the pipeline times out before a user sends signal to retry, the pipeline results in timeout. 

> It sounds like the Pipeline would be in a suspended kind of state whenever at least a PipelineTask that opted-in for on-demand-retry fails.

Yes, the pipeline will be in a suspended state. 

> Would the user be able to cancel a pipeline in such state?
Yes, the cancellation works as is, the user will be able to cancel a pipeline in suspended state.

> Would the user be able to let the pipeline fail if they decided to?
We can discuss this further. I was trying to keep this simple and support only allow to restart but we can definitely add support for declaring a pipeline fail if needed.

> What happens is resources (like the Pod) are pruned while waiting for a user to request a retry?
A new pod is created very similar to the existing retry functionality where the tekton controller creates a new pod for every new retry, if I am not mistaken 🤔